autosklearn/ensemble_builder.py

```diff
@@ -857,15 +890,23 @@ def predict(self, set_: str,
         if self.SAVE2DISC:
             self.backend.save_ensemble(ensemble, index_run, self.seed)
 
+        if set_ == 'valid':
+            pred_set = Y_VALID
+        elif set_ == 'test':
+            pred_set = Y_TEST
+        else:
+            pred_set = Y_ENSEMBLE
         predictions = np.array([
-            self.read_preds[k][Y_VALID if set_ == 'valid' else Y_TEST]
+            self.read_preds[k][pred_set]
             for k in selected_keys
         ])
 
         if n_preds == predictions.shape[0]:
             y = ensemble.predict(predictions)
             if self.task_type == BINARY_CLASSIFICATION:
-                y = y[:, 1]
+                # If on a binary task, and uncertain predictions (0.5)
```

Comment:
This looks dangerous and I don't understand the reason for it. Are we doing this somewhere else, too?

Reply:
Yeah, this is the equivalent of https://github.com/automl/auto-sklearn/blob/5126d52c682b0ec32fa21eff9d568cee9cda0bd2/autosklearn/metrics/__init__.py#L59

It was not seen before, because we never utilized the training predictions of the ensemble (just test/validation).

This one is thought to reproduce. It happened while testing complicated datasets where the output was 0.5, of the individual models, and then the ensemble mimic that. I added this as a safety measure, again because we do that in L59. The problem is that 0.5 is considered not a binary output, rather continuous, by check-targets and hence I rely in BINARY_CLASSIFICATION to score the prediction for display.


This prediction is just used for tracking. But, if we are dealing with binary classification, aren't you bound to do the np.argmax?

Reply:
In case of area under the curve, the prediction needs to be the probability.

Reply:
Thanks for pointing this out.

We do need a mechanism to handle 0.5 probabilities, else we run into:

```
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/home/chico/work/auto-sklearn_fork_ensemble_history/venv/lib/python3.6/site-packages/pynisher/limit_function_call.py", line 93, in subprocess_func
    return_value = ((func(*args, **kwargs), 0))
  File "/home/chico/work/auto-sklearn_fork_ensemble_history/venv/lib/python3.6/site-packages/auto_sklearn-0.8.0-py3.6.egg/autosklearn/ensemble_builder.py", line 364, in main
    test_pred
  File "/home/chico/work/auto-sklearn_fork_ensemble_history/venv/lib/python3.6/site-packages/auto_sklearn-0.8.0-py3.6.egg/autosklearn/ensemble_builder.py", line 970, in _add_ensemble_trajectory
    all_scoring_functions=False
  File "/home/chico/work/auto-sklearn_fork_ensemble_history/venv/lib/python3.6/site-packages/auto_sklearn-0.8.0-py3.6.egg/autosklearn/metrics/__init__.py", line 312, in calculate_score
    score = metric(solution, prediction)
  File "/home/chico/work/auto-sklearn_fork_ensemble_history/venv/lib/python3.6/site-packages/auto_sklearn-0.8.0-py3.6.egg/autosklearn/metrics/__init__.py", line 74, in __call__
    **self._kwargs)
  File "/home/chico/work/auto-sklearn_fork_ensemble_history/venv/lib/python3.6/site-packages/sklearn/metrics/_classification.py", line 185, in accuracy_score
    y_type, y_true, y_pred = _check_targets(y_true, y_pred)
  File "/home/chico/work/auto-sklearn_fork_ensemble_history/venv/lib/python3.6/site-packages/sklearn/metrics/_classification.py", line 90, in _check_targets
    "and {1} targets".format(type_true, type_pred))
ValueError: Classification metrics can't handle a mix of binary and continuous targets
```

I think the problem is due to the following:

1. Ensemble builder saves predictions with y[,:1] so that it always assume we want 1 dim prediction
2. nevertheless, the predict scorer assumes you have 2d predictions for classification (look at https://github.com/automl/auto-sklearn/blob/859fb993607bd5cfb881aca3b346cf4d1addf470/autosklearn/metrics/__init__.py#L53)
3. There are 2 possible solutions. Change the ensemble building saving process (and costumers) not to expect a y[:,1] prediction, or enhance the predictscorer to handler this scenario.

I went with the second approach. Kindly let me know your thoughts about this issue. Do recall that we have seen this before with the balanced accuracy, where a continuous to binary prediction caused problems.

Reply:
This looks good to me - unit tests are working as well, so I'm going to merge this one.